lib/itertools/stream/stream.go

```diff
@@ -0,0 +1,459 @@
+	if size < 1 {
+		return Empty[[]T]()
+	}
+	return func(yield func([]T, error) bool) {
+		var chunk []T
+		for item, err := range stream {
+			if err != nil {
+				yield(nil, trace.Wrap(err))
+				return
+			}
+
+			if chunk == nil {
+				chunk = make([]T, 0, size)
+			}
+
+			chunk = append(chunk, item)
+			if len(chunk) == size {
+				if !yield(chunk, nil) {
+					return
+				}
+				chunk = nil
+			}
+		}
+
+		if len(chunk) > 0 {
+			if !yield(chunk, nil) {
+				return
+			}
+		}
+	}
+}
+
+// Fail creates an empty stream that fails immediately with the supplied error.
+func Fail[T any](err error) Stream[T] {
+	return func(yield func(T, error) bool) {
+		if err != nil {
+			var zero T
+			yield(zero, trace.Wrap(err))
+		}
+	}
+}
+
+// Empty creates an empty stream.
+func Empty[T any]() Stream[T] {
+	return func(yield func(T, error) bool) {}
+}
+
+// Once creates a stream that yields a single item.
+func Once[T any](item T) Stream[T] {
+	return func(yield func(T, error) bool) {
+		yield(item, nil)
+	}
+}
+
+// OnceFunc builds a stream from a closure that will yield exactly zero or one items. This stream
+// is the lazy equivalent of the Once/Fail/Empty combinators. A nil error value results
+// in a single-element stream. An error value of io.EOF results in an empty stream. All other error
+// values result in a failing stream.
+func OnceFunc[T any](fn func() (T, error)) Stream[T] {
+	return func(yield func(T, error) bool) {
+		item, err := fn()
+		if errors.Is(err, io.EOF) {
+			return
+		}
+		yield(item, err)
+	}
+}
+
+// Drain consumes a stream to completion, reporting its error if any.
+func Drain[T any](stream Stream[T]) error {
+	for _, err := range stream {
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Slice constructs a stream from a slice.
+func Slice[T any](items []T) Stream[T] {
+	return func(yield func(T, error) bool) {
+		for _, item := range items {
+			if !yield(item, nil) {
+				return
+			}
+		}
+	}
+}
+
+// PageFunc is equivalent to Func except that it performs internal depagination. As with
+// Func, the supplied closure *must* return io.EOF if no more items are available. Failure
+// to return io.EOF (or some other error) may result in infinite loops.
+func PageFunc[T any](fn func() ([]T, error)) Stream[T] {
+	return func(yield func(T, error) bool) {
+		for {
+			items, err := fn()
+			if err != nil {
+				if !errors.Is(err, io.EOF) {
+					var zero T
+					yield(zero, trace.Wrap(err))
+				}
+				return
+			}
+
+			for _, item := range items {
+				if !yield(item, nil) {
+					return
+				}
+			}
+		}
+	}
+}
+
+// Skip skips the first n items from a stream. Zero/negative values of n
+// have no effect.
+func Skip[T any](stream Stream[T], n int) Stream[T] {
+	return func(yield func(T, error) bool) {
+		for item, err := range stream {
+			if err != nil {
+				var zero T
+				yield(zero, trace.Wrap(err))
+				return
+			}
+
+			if n > 0 {
+				n--
```

Comment:
n is shared for the whole `Stream`, it should be copied and it should be local to the execution of the stream instead IMO